gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/v4/TcpApiReactor.java

```diff
@@ -0,0 +1,303 @@
+        EndpointManager endpointManager,
+        RequestTimeoutConfiguration requestTimeoutConfiguration
+    ) {
+        this.api = api;
+        this.node = node;
+        this.requestTimeoutConfiguration = requestTimeoutConfiguration;
+        this.entrypointConnectorResolver =
+            new DefaultEntrypointConnectorResolver(api.getDefinition(), deploymentContext, entrypointConnectorPluginManager);
+        this.endpointManager = endpointManager;
+        this.defaultInvoker = new EndpointInvoker(endpointManager);
+        this.tracingEnabled = configuration.getProperty(SERVICES_TRACING_ENABLED_PROPERTY, Boolean.class, false);
+        this.pendingRequestsTimeout = configuration.getProperty(PENDING_REQUESTS_TIMEOUT_PROPERTY, Long.class, 10_000L);
+        this.lifecycleState = Lifecycle.State.INITIALIZED;
+        this.loggingExcludedResponseType =
+            configuration.getProperty(REPORTERS_LOGGING_EXCLUDED_RESPONSE_TYPES_PROPERTY, String.class, null);
+        this.loggingMaxSize = configuration.getProperty(REPORTERS_LOGGING_MAX_SIZE_PROPERTY, String.class, null);
+    }
+
+    @Override
+    public Api api() {
+        return this.api;
+    }
+
+    @Override
+    public Completable handle(MutableExecutionContext ctx) {
+        // might need a bit of refactoring (duplication)
+        ctx.setAttribute(ContextAttributes.ATTR_API, api.getId());
+        ctx.setAttribute(ContextAttributes.ATTR_API_DEPLOYED_AT, api.getDeployedAt().getTime());
+        ctx.setAttribute(ContextAttributes.ATTR_ORGANIZATION, api.getOrganizationId());
+        ctx.setAttribute(ContextAttributes.ATTR_ENVIRONMENT, api.getEnvironmentId());
+        ctx.setInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER, defaultInvoker);
+
+        // TODO specific Tcp API Request processor chain factory that contains SubscriptionProcessor in beforeApi chain
+
+        return new CompletableReactorChain(handleEntrypointRequest(ctx))
+            // configure backend call
+            .chainWith(invokeBackend(ctx))
+            // setup timeout
+            .chainWith(upstream -> timeout(upstream, ctx))
+            // handle response
+            .chainWith(handleEntrypointResponse(ctx))
+            // hook everything up and start piping data
+            .chainWith(ctx.response().end(ctx))
+            .doOnSubscribe(disposable -> pendingRequests.incrementAndGet())
+            .doFinally(pendingRequests::decrementAndGet);
+    }
+
+    protected Completable handleEntrypointRequest(final MutableExecutionContext ctx) {
+        // CAUTION, code is duplicated with ApiReactor
+        return Completable.defer(() -> {
+            final EntrypointConnector entrypointConnector = entrypointConnectorResolver.resolve(ctx);
+            if (entrypointConnector == null) {
+                return ctx.interruptWith(
+                    new ExecutionFailure(HttpStatusCode.NOT_FOUND_404).message("No entrypoint matches the incoming request")
```

Comment:
should we use a `TcpCode` enum to manage our own list of status for a tcp request ?
@jhaeyaert @benoitgravitee

Reply:
I think it makes no sense to have code here as we don't know what the protocol is. Does a `TcpExecutionFailure` making no use of code make sense instead?

Reply:
I can add it to gateway-api :)